### PR TITLE
Move new buffer allocation out of lock

### PR DESCRIPTION
**Pull request by benaadams**
Slim lock scope as suggested in  #7727

/cc @stephentoub

### CODE REVIEW

File: src/System.Buffers/src/System/Buffers/DefaultArrayPoolBucket.cs [hunk 1]
```diff
@@ -55,14 +56,7 @@ internal T[] Rent()
...
                     {
-                        buffer = new T[_bufferLength];
-                        if (ArrayPoolEventSource.Log.IsEnabled())
-                            ArrayPoolEventSource.Log.BufferAllocated(
-                                Utilities.GetBufferId(buffer),
-                                _bufferLength,
-                                _poolId,
-                                Utilities.GetBucketId(this),
-                                ArrayPoolEventSource.BufferAllocationReason.Pooled);
+                        alloc = true;
```
Comment (jasonwilliams200OK) [inline src/System.Buffers/src/System/Buffers/DefaultArrayPoolBucket.cs]: nit: `alloc = buffer == null;` to get rid of the branch/nesting.

Comment (benaadams) [inline src/System.Buffers/src/System/Buffers/DefaultArrayPoolBucket.cs]: done

**Comment by stephentoub** [comment]
Thanks, Ben.  I sent several variations of this last night to Ryan for him to try out.  We can also consider changing the implementation more significantly, but before doing that we'd want to really understand where time is being spent and why, what kind of contention is being encountered, etc.

**Comment by benaadams** [comment]
`[ThreadStatic]` would reduce the contention, but likely allocate too much. 

Could round robin?
e.g. 2^N buffer pools round robining on an `Interlocked.Increment(ref poolIndex) | (2^N - 1)` - should only be cache contention on poolIndex rather than on the lock.

**Comment by stephentoub** [comment]
There are many different ways a pool can be implemented, with many different techniques that each incur different trade-offs that benefit and hurt different situations.  We don't actually know that there's a problem yet, and if there is one, we can't reasonably re-design it without more detailed information to know what the actual issues are.

**Comment by stephentoub** [comment]
Regardless, since the only extra cost in this change is one extra comparison, and in exchange we move allocations out of the region, it should be pure goodness, so we can go ahead with it.  If it doesn't move the needle (and if the needle needs to be moved), we can look at alternatives.

**Pull request merged**